cmd/finch/nerdctl.go

```diff
@@ -142,95 +143,202 @@ func (nc *nerdctlCommand) run(cmdName string, args []string) error {
-			b, _, _ := strings.Cut(arg, "=")
-			h, ok := aMap[b]
-			if ok {
-				err = h(nc.systemDeps, args, i)
-				if err != nil {
-					return err
-				}
-				// This is required when the positional argument at i is mutated by argHandler, eg -v=C:\Users:/tmp:ro
-				arg = args[i]
-			}
-		}
-		// parsing environment values from the command line may pre-fetch and
-		// consume the next argument; this loop variable will skip these pre-consumed
-		// entries from the command line
-		if skip {
-			skip = false
-			continue
-		}
+	switch cmdName {
+	case "container", "run", "exec", "compose", "container run":
+		// check if an option flag is present; immediately following the command
 		switch {
-		case arg == "--debug":
-			// explicitly setting log level to avoid `--debug` flag being interpreted as nerdctl command
-			if envArgPos == -1 {
-				isDebug = 1
-			}
-			nc.logger.SetLevel(flog.Debug)
-		case argIsEnv(arg):
-			if envArgPos == -1 {
-				envArgPos = i - isDebug
+		case args[0] == "run" && strings.HasPrefix(args[1], "-"):
+			firstOptPos = 1
+		case strings.HasPrefix(args[0], "-"):
+			firstOptPos = 0
+		default:
+			firstOptPos = -1
+		}
+		// lastOpt is used to track when no more flags need to be processed
+		lastOpt = false
+
+		shortFlagBoolSet := sets.NewString("-d", "-i", "-t", "-di", "-dt", "-it", "-dit")
+		longFlagBoolSet := sets.NewString(
+			"--detach", "--init", "--interactive", "--oom-kill-disable",
+			"--privileged", "--read-only", "--rm", "--rootfs", "--tty")
+
+		shortFlagArgsSet := sets.NewString("-h", "-m", "-u", "-w", "-p", "-l", "-v")
+
+		for i, arg := range args {
+			// Check if command requires arg handling
+			if hasArgHandler {
+				// Check if argument for the command needs handling, sometimes it can be --file=<filename>
+				b, _, _ := strings.Cut(arg, "=")
+				h, ok := aMap[b]
+				if ok {
+					err = h(nc.systemDeps, args, i)
+					if err != nil {
+						return err
+					}
+					// This is required when the positional argument at i is mutated by argHandler, eg -v=C:\Users:/tmp:ro
+					arg = args[i]
+				}
 			}
-			shouldSkip, addEnv := handleEnv(nc.systemDeps, arg, args[i+1])
-			skip = shouldSkip
-			if addEnv != "" {
-				envs = append(envs, addEnv)
+			// parsing arguments from the command line
+			// may pre-fetch and consume the next argument;
+			// the loop variable will skip any pre-consumed args
+			if skip {
+				skip = false
+				continue
 			}
-		case strings.HasPrefix(arg, "--env-file"):
-			if envArgPos == -1 {
-				envArgPos = i - isDebug
+
+			// after last Option position is found, pass through each arg as an internal command argument (short-circuit)
+			if lastOpt {
+				cmdArgs = append(cmdArgs, arg)
+				continue
 			}
 
-			shouldSkip, addEnvs, err := handleEnvFile(nc.fs, nc.systemDeps, arg, args[i+1])
-			if err != nil {
-				return err
+			// scan the arg to separate shortFlag combinations of boolFlag and argFlag
+			re := regexp.MustCompile(`^-[a-z^-]+[ehmuwplv]$`)
```

Comment:
If nerdctl adds a new bool/short flag, do we then have to update this regexp?

Reply:
This regex, `shortFlagBoolSet` (line 160), the `longFlagBoolSet` (line 161-164), and the `shortFlagArgsSet` (in line165) are all susceptible to changes with the short flags.

To be honest, this regex was a "proof of concept" for handling the qty.3 enforcement goals in one line.
However, I would be easily convinced to refactor this to use the clearly defined sets (lines 160-165).

Reply:
While refactoring this code, I also made an improvement for dynamically handling miscellaneous shortFlagBool combinations.